src/protocol/malicious.rs

```diff
@@ -0,0 +1,350 @@
+        // This is a bit tricky, because we cannot count on the multiplications being executed in the same order across all the helpers.
+        // The easiest way is to just narrow the context used to perform the multiplication, and then re-use the same record_id.
+        // This ensures that when the helpers all go to validate the multiplication: "1/foo/bar/baz", they all use the prss from "1/foo/bar/baz/validate".
+        // That way, we don't need to worry about the order in which the multiplications are executed.
+        let random_constant = ctx.prss().generate_replicated(record_id);
+
+        self.u += Self::compute_dot_product_contribution(random_constant, rx);
+        self.w += Self::compute_dot_product_contribution(random_constant, x);
+    }
+
+    #[allow(dead_code)]
+    pub async fn validate(&self) -> Result<(), BoxError> {
+        let record_3 = RecordId::from(3);
+        let record_4 = RecordId::from(4);
+        let record_5 = RecordId::from(5);
+        let record_6 = RecordId::from(6);
+
+        // send our `u_i+1` value to the helper on the right
+        let channel = self.ctx.mesh();
+        try_join(
+            channel.send(
+                self.ctx.role().peer(Direction::Right),
+                record_3,
+                UValue { payload: self.u },
+            ),
+            channel.send(
+                self.ctx.role().peer(Direction::Right),
+                record_4,
+                UValue { payload: self.w },
+            ),
+        )
+        .await?;
+
+        // receive `u_i` value from helper to the left
+        let (u_left_struct, w_left_struct): (UValue<F>, UValue<F>) = try_join(
+            channel.receive(self.ctx.role().peer(Direction::Left), record_3),
+            channel.receive(self.ctx.role().peer(Direction::Left), record_4),
+        )
+        .await?;
+
+        let u_left = u_left_struct.payload;
+        let w_left = w_left_struct.payload;
+
+        let u_share = Replicated::new(u_left, self.u);
+        let w_share = Replicated::new(w_left, self.w);
+
+        // This should probably be done in parallel with the futures above
+        let r = reveal(self.ctx.narrow(&Step::RevealR), record_5, self.r_share).await?;
+        let t = u_share - (w_share * r);
+
+        let is_valid = check_zero(self.ctx.narrow(&Step::CheckZero), record_6, t).await?;
+
+        if is_valid {
+            Ok(())
+        } else {
+            Err(Box::new(Error::MaliciousSecurityCheckFailed))
+        }
+    }
+}
+
+#[cfg(test)]
+pub mod tests {
+    use crate::error::BoxError;
+    use crate::field::Fp31;
+    use crate::protocol::{
+        malicious::{SecurityValidator, Step},
+        QueryId, RecordId,
+    };
+    use crate::test_fixture::{logging, make_contexts, make_world, share, TestWorld};
+    use futures::future::{try_join, try_join_all};
+    use proptest::prelude::Rng;
+
+    #[tokio::test]
```

Comment:
Some commentary here might help establish what the goal of the test is.  It's easy to follow, but...